src/main/java/org/wildfly/security/authz/MappedRoleMapper.java

```diff
@@ -0,0 +1,171 @@
+ *
+ * @author <a href="mailto:mmazanek@redhat.com">Martin Mazanek</a>
+ */
+public class MappedRoleMapper implements RoleMapper {
+
+    private final Map<String, Set<String>> reverseRoleMap;
+    private final boolean keepMapped;
+    private final boolean keepNonMapped;
+    private final Roles keyRoles;
+
+    /**
+     * Construct a new instance.
+     * Called from WildFly core when using this as a custom component. You should not use this constructor and use {@link MappedRoleMapper.Builder} instead.
+     *
+     * @see MappedRoleMapper.Builder
+     */
+    public MappedRoleMapper() {
+        this.reverseRoleMap = new LinkedHashMap<>();
+        this.keepMapped = false; //not keeping any delegate roles as there is no way to avoid configuration conflicts
+        this.keepNonMapped = false; //this functionality is possible to recreate by manually combining role mappers
+        this.keyRoles = Roles.NONE;
+    }
+
+
+    private MappedRoleMapper(Map<String, Set<String>> roleMap, boolean keepMapped, boolean keepNonMapped) {
+        this.reverseRoleMap = new LinkedHashMap<>();
+        this.keepMapped = keepMapped;
+        this.keepNonMapped = keepNonMapped;
+        this.keyRoles = Roles.fromSet(roleMap.keySet());
+
+        Set<Map.Entry<String, Set<String>>> entrySet = roleMap.entrySet();
+
+        for (Map.Entry<String, Set<String>> entry : entrySet) {
+
+            for (String mappedRole : entry.getValue()) {
+                Set<String> rolesToMappedRole = reverseRoleMap.get(mappedRole);
+
+                if (rolesToMappedRole == null) {
+                    rolesToMappedRole = new LinkedHashSet<>();
+                    reverseRoleMap.put(mappedRole, rolesToMappedRole);
+                }
+
+                rolesToMappedRole.add(entry.getKey());
+            }
+        }
+    }
+
+
+    /**
+     * Custom component method.
+     * Called from WildFly core. Used to include mapped role mapping functionality in older WildFly versions.
+     *
+     * @param configuration map of mapping rules
+     */
+    public void initialize(final Map<String, String> configuration) {
+        configuration.forEach( (key, value) -> {
+            String[] newRoles = value.split(",");
+            for (String newRole : newRoles) {
+                Set<String> rolesToMappedRole = reverseRoleMap.get(newRole);
+                if (rolesToMappedRole == null) {
+                    rolesToMappedRole = new LinkedHashSet<>();
+                    reverseRoleMap.put(newRole, rolesToMappedRole);
+                }
+                rolesToMappedRole.add(value);
+            }
+        } );
+    }
+
+    @Override
+    public Roles mapRoles(Roles rolesToMap) {
+        Roles res = new MappedRoles(rolesToMap, this.reverseRoleMap);
+        if (keepMapped) {
+            res = res.or(rolesToMap.and(keyRoles));
+        }
+        if (keepNonMapped) {
+            res = res.or(rolesToMap.minus(keyRoles));
+        }
```

Comment:
I agree with the keeping it as simple as we can but as we found when we tried to use logical permission mappers end users found it complex to understand - if we can offer a simplification to configuration that is usable via a single boolean that may also be beneficial.

Reply:
Ok, but would not be this in subsystem sufficient, to keep at least mapper class simple?
(booleans would be adding logical structure above in the subsystem)

Reply:
+1 If we can assemble it in the subsystem automatically we can probably simplify the implementation within Elytron as far as possible.

Reply:
In that case, can I remove the Builder class aswell, as there is only one configurable attribute - the actual map? It would also simplify the issue with having public constructor just for being able to use the mapper in older AS as a custom component.

Reply:
One thing we could however do with the builder is add the mappings one at a time.

Reply:
Alright, so I am keeping it

Reply:
Yeah. In situation when we face complaints about Elytron being difficult to configure I would prefer to provide boolean parameter instead of recommendation to use logical resource.

I personally can imagine all four permutation can be handy. Note, in legacy we have one boolean / two permutation [1] - basically "keep mapped" alternative.  

[1] https://access.redhat.com/documentation/en-us/red_hat_jboss_enterprise_application_platform/7.0/html-single/login_module_reference/#rolemapping_login_module